File: recursion/javascript/solutions/linkedList.js
/**
 * Shared linked list infrastructure for solutions.
 */

const util = require('node:util');

const EMPTY_LIST = null;

class Node {
  static fromArray(array) {
    if (array.length === 0) {
      return EMPTY_LIST;
    }
    let [first, ...rest] = array;
    return prepend(first, Node.fromArray(rest));
  }

  constructor(value, next = EMPTY_LIST) {
    this.value = value;
    this.next = next;
  }

  toString() {
    return listToString(this);
  }

  [util.inspect.custom]() {
    return this.toString();
  }
}

function listToString(list) {
  if (isEmpty(list)) {
    return '()';
  }
  let [first, rest] = unprepend(list);
  return `${first} -> ${listToString(rest)}`;
}

function isEmpty(list) {
  return list === EMPTY_LIST;
}

function prepend(val, list) {
  return new Node(val, list);
}

function unprepend(list) {
  return [list.value, list.next];
}

function toArray(list) {
  if (isEmpty(list)) {
    return [];
  }
  let [first, rest] = unprepend(list);
  return [first, ...toArray(rest)];
}

module.exports = {
  EMPTY_LIST,
  Node,
  listToString,
  isEmpty,
  prepend,
  unprepend,
  toArray,
};

File: recursion/python/solutions/linked_list.py
"""Shared linked list infrastructure for solutions."""

EMPTY_LIST = None


class Node:
    """A node in a singly-linked list."""

    @staticmethod
    def from_list(lst):
        """Create a linked list from a Python list."""
        if len(lst) == 0:
            return EMPTY_LIST
        first, *rest = lst
        return prepend(first, Node.from_list(rest))

    def __init__(self, value, next=None):
        self.value = value
        self.next = next

    def __str__(self):
        return list_to_string(self)

    def __repr__(self):
        return self.__str__()


def list_to_string(lst):
    """Convert linked list to string representation."""
    if is_empty(lst):
        return '()'

    first, rest = unprepend(lst)

    return f'{first} -> {list_to_string(rest)}'


def is_empty(lst):
    """Return True if lst is the empty list."""
    return lst is EMPTY_LIST


def prepend(val, lst):
    """Add val to the front of lst."""
    return Node(val, lst)


def unprepend(lst):
    """Return (first element, rest of list) tuple."""
    return lst.value, lst.next


def to_python_list(lst):
    """Convert linked list to Python list for easier testing."""
    if is_empty(lst):
        return []
    first, rest = unprepend(lst)
    return [first] + to_python_list(rest)
